### PR TITLE
feat(duplicated_node_checker): add packages to check duplication of node names in ros2

### PR DESCRIPTION
**Pull request by Owen-Liuyuxuan**
## Description

Create a new package of duplicated_node_checker. The goal is to detect duplication of node names in the ROS2 space and publish diagnostic messages.

## Related links
https://github.com/autowarefoundation/autoware_launch/pull/631

## Tests performed

PSim

After also adjusting parameters in the autoware_launch package, it is possible to get the notice from the mrmsummary plugin (tier4_system_rviz_plugin) 

Screenshot from 2023-10-12 15-40-06

Even if we only launch the diagnostic aggregator, we can also check the diagnostic values from the `ros2 run rqt_robot_monitor rqt_robot_monitor`
image

## Notes for reviewers

For now, the resulting state is only either "OK" or "ERROR", and it is enough for monitoring. If this is fine, I will add this parameter to the autoware_launch and make a new PR there.

## Interface changes

The node only requires a `update_rate` parameter. Notice that this is highly related to how system_error_monitor tagging data staled.

## Effects on system behavior

In terms of computation consumption, this PR uses std::map for duplication detection, so the time/space consumption will both be O(n) w.r.t. to the number of nodes (generally we have 100+ in Autoware), thus, it is manageable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: common/duplicated_node_checker/include/duplicated_node_checker/duplicated_node_checker_core.hpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// Copyright 2023 Tier IV, Inc.
```
Comment (kminoda) [inline common/duplicated_node_checker/include/duplicated_node_checker/duplicated_node_checker_core.hpp:1]: nits:
```suggestion
// Copyright 2023 TIER IV, Inc.
```

**Review by kminoda — COMMENTED** [review]
Thank you for the PR! Awesome work :tada: 
Please change the config in autoware_launch as well :pray:

File: common/duplicated_node_checker/include/duplicated_node_checker/duplicated_node_checker_core.hpp (hunk 1 above)
Comment (Owen-Liuyuxuan) [reply common/duplicated_node_checker/include/duplicated_node_checker/duplicated_node_checker_core.hpp:1]: I saw other packages like system_error_monitor use "Tier IV" in the license. Did something changed along the way?

**Comment by Owen-Liuyuxuan** [comment]
Corresponding PR in launch package. 

https://github.com/autowarefoundation/autoware_launch/pull/631

**Comment by shmpwk** [comment]
@ito-san 
Do you think this duplicated node checker can be launched in system launch or standalone as it is?

**Comment by ito-san** [comment]
> @ito-san Do you think this duplicated node checker can be launched in system launch or standalone as it is?

@shmpwk 
I believe it should be fine if it doesn't affect other nodes that will be launched together.

**Comment by shmpwk** [comment]
@ito-san 
You mean this node can be launched in system.launch.xml ? Any other thoughts?
https://github.com/autowarefoundation/autoware.universe/blob/main/launch/tier4_system_launch/launch/system.launch.xml

**Comment by Owen-Liuyuxuan** [comment]
> In that case, IMO this packages is better located at system module rather than common, is it OK?

@shmpwk @ito-san 
From my understanding of the organization logic of the common module and the system module, I also agreed that this package should be in the system module.

**Comment by ito-san** [comment]
@shmpwk @Owen-Liuyuxuan 
I agree this package should be located in the system If user can use Autoware without any issues in a standard manner.

**Comment by shmpwk** [comment]
@Owen-Liuyuxuan 
Could you launch this in 
 https://github.com/autowarefoundation/autoware.universe/blob/main/launch/tier4_system_launch/launch/system.launch.xml

**Comment by tkimura4** [comment]
@Owen-Liuyuxuan 
Could you also create a pull reqeust for autoware_launch?

https://github.com/autowarefoundation/autoware_launch/tree/main/autoware_launch/config/system/system_error_monitor

**Comment by tkimura4** [comment]
Could you also create a pull request for autoware_launch?
https://github.com/autowarefoundation/autoware_launch/tree/main/autoware_launch/config/system/system_error_monitor

**Comment by Owen-Liuyuxuan** [comment]
@tkimura4 

> Could you also create a pull request for autoware_launch?

I have a PR on https://github.com/autowarefoundation/autoware_launch/pull/631 
I will test the launch and parameters on each side. And also try to follow the JSON schema. I may need some time to test the path and so on.

**Review by shmpwk — APPROVED** [review]
Excellent!

**Comment by shmpwk** [comment]
@ito-san @asana17 
Could you just approve this as a code owner? I don't have right to merge this PR.

**Review by ito-san — APPROVED** [review]
LGTM

**Pull request merged**